display_list/display_list_path_effect.h

```diff
@@ -0,0 +1,155 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_DISPLAY_LIST_DISPLAY_LIST_PATH_EFFECT_H_
+#define FLUTTER_DISPLAY_LIST_DISPLAY_LIST_PATH_EFFECT_H_
+
+#include "flutter/display_list/display_list_attributes.h"
+#include "flutter/display_list/types.h"
+#include "flutter/fml/logging.h"
+#include "include/core/SkScalar.h"
+
+namespace flutter {
+
+class DlDashPathEffect;
+
+// The DisplayList PathEffect class. This class implements all of the
+// facilities and adheres to the design goals of the |DlAttribute| base
+// class.
+
+// An enumerated type for the recognized PathEffect operations.
+// In current Flutter we only use the DashPathEffect.
+// And another PathEffect outside of the recognized types is needed
+// then a |kUnknown| type that simply defers to an SkPathEffect is
+// provided as a fallback.
+enum class DlPathEffectType {
+  kDash,
+  kUnknown,
+};
+
+class DlPathEffect
+    : public DlAttribute<DlPathEffect, SkPathEffect, DlPathEffectType> {
+ public:
+  static std::shared_ptr<DlPathEffect> From(SkPathEffect* sk_path_effect);
+
+  static std::shared_ptr<DlPathEffect> From(
+      sk_sp<SkPathEffect> sk_path_effect) {
+    return From(sk_path_effect.get());
+  }
+
+  virtual const DlDashPathEffect* asDash() const { return nullptr; }
+};
+
+/// The DashPathEffect which specifies modifying the path effect, and it
+/// only affects storked paths.
+/// intervals: array containing an even number of entries (>=2), with
+/// the even indices specifying the length of "on" intervals, and the odd
+/// indices specifying the length of "off" intervals. This array will be
+/// copied in Make, and can be disposed of freely after.
+/// count: number of elements in the intervals array
+/// phase: offset into the intervals array (mod the sum of all of the
+/// intervals).
+///
+/// For example: if intervals[] = {10, 20}, count = 2, and phase = 25,
+/// this will set up a dashed path like so:
+/// 5 pixels off
+/// 10 pixels on
+/// 20 pixels off
+/// 10 pixels on
+/// 20 pixels off
+/// ...
+/// A phase of -5, 25, 55, 85, etc. would all result in the same path,
+/// because the sum of all the intervals is 30.
+///
+class DlDashPathEffect : public DlPathEffect {
+ public:
+  DlDashPathEffect(const SkScalar intervals[], int count, SkScalar phase)
+      : count_(count), phase_(phase) {
+    intervals_ = new SkScalar[count];
+    if (intervals != nullptr) {
+      memcpy(intervals_, intervals, sizeof(SkScalar) * count);
```

Comment:
If you switch to a system that allocates the storage for the intervals in line with the object then DisplayList::Equals can do a bulk compare. Otherwise you'll have to interrupt the bulk compares to do a manual equals in the associated SetPathEffectOp.

See DlColorSource::MakeLinear for an example of how that works.

Reply:
I have change the way of allocates memory for `intervals_`. Is this what you want?